torch/optim/nadam.py

```diff
@@ -116,23 +122,145 @@ def step(self, closure=None):
+        func = _multi_tensor_nadam
+    else:
+        func = _single_tensor_nadam
+
+    func(params,
+         grads,
+         exp_avgs,
+         exp_avg_sqs,
+         mu_products,
+         state_steps,
+         beta1=beta1,
+         beta2=beta2,
+         lr=lr,
+         weight_decay=weight_decay,
+         momentum_decay=momentum_decay,
+         eps=eps)
+
+
+def _single_tensor_nadam(params: List[Tensor],
+                         grads: List[Tensor],
+                         exp_avgs: List[Tensor],
+                         exp_avg_sqs: List[Tensor],
+                         mu_products: List[float],
+                         state_steps: List[int],
+                         *,
+                         beta1: float,
+                         beta2: float,
+                         lr: float,
+                         weight_decay: float,
+                         momentum_decay: float,
+                         eps: float):
+
+    for i, param in enumerate(params):
+        grad = grads[i]
+        exp_avg = exp_avgs[i]
+        exp_avg_sq = exp_avg_sqs[i]
+        mu_product = mu_products[i]
+        step = state_steps[i]
+
+        bias_correction2 = 1 - beta2 ** step
+
+        if weight_decay != 0:
+            grad = grad.add(param, alpha=weight_decay)
+
+        # calculate the momentum cache \mu^{t} and \mu^{t+1}
+        mu = beta1 * (1. - 0.5 * (0.96 ** (step * momentum_decay)))
+        mu_next = beta1 * (1. - 0.5 * (0.96 ** ((step + 1) * momentum_decay)))
+        mu_product = mu_product * mu
+        mu_product_next = mu_product * mu * mu_next
+
+        # decay the first and second moment running average coefficient
+        exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+        exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)
+
+        denom = exp_avg_sq.div(bias_correction2).sqrt().add_(eps)
+        param.addcdiv_(grad, denom, value=-lr * (1. - mu) / (1. - mu_product))
+        param.addcdiv_(exp_avg, denom, value=-lr * mu_next / (1. - mu_product_next))
+
+
+def _multi_tensor_nadam(params: List[Tensor],
+                        grads: List[Tensor],
+                        exp_avgs: List[Tensor],
+                        exp_avg_sqs: List[Tensor],
+                        mu_products: List[float],
+                        state_steps: List[int],
+                        *,
+                        beta1: float,
+                        beta2: float,
+                        lr: float,
+                        weight_decay: float,
+                        momentum_decay: float,
+                        eps: float):
+
+    bias_correction2 = [1 - beta2 ** step for step in state_steps]
+    mus = [beta1 * (1. - 0.5 * (0.96 ** (step * momentum_decay))) for step in state_steps]
+    mu_nexts = [beta1 * (1. - 0.5 * (0.96 ** ((step + 1) * momentum_decay)))
+                for step in state_steps]
+    mu_products = [mu * mu_product for mu, mu_product in zip(mus, mu_products)]
```

Comment:
The multitensor Nadam class does [an update on mu_product](https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/nadam.py#L90-L92) before the call to nadam and then does computations on `mu_products` directly whereas the singletensor class does [an update on mu_product](https://github.com/pytorch/pytorch/blob/master/torch/optim/nadam.py#L132-L136) after the call to nadam and computes the new `mu_products` within the nadam function. In this combined class I preserved the singletensor behavior so I added line 245 here to ensure the same behavior

Reply:
This could be done with a foreach op no?
Also the existing code is modifying mu_products inplace, don't we want to preserve that/

Reply:
This could be done with a foreach op no?
Also the existing code is modifying mu_products inplace, don't we want to preserve that?

Reply:
Hm I didn't do this in place because each `mu_product` is a float rather than a tensor so even if it's done in-place i don't think it updates the underlying `state['mu_product']`, but there is an update step within the NAdam.step function on line 146 which updates `state['mu_product']`, which is preserved from the single tensor version, am I thinking about this correctly?

Reply:
Aren't you computing the same thing twice here then?
I am a bit confused now where this value is updated for each case.

In general, I think we want to stay as close to the original code as possible. Even if we have to fold some of the state update in the functional function.

Reply:
Yes I agree that the same thing is being computed twice, but that was what was being done initially in torch/optim/nadam.py. The single tensor version does the update of mu_product [after calling F.nadam](https://github.com/pytorch/pytorch/blob/master/torch/optim/nadam.py#L132-L136) (and this same computation is done in the function on lines 224 and 226) whereas the multitensor one does it [before calling F.nadam](https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/nadam.py#L90-L92) and so the logic isn't repeated in the function! So I think we need to change the code of either one of the functional forms (for example if I preserve the initial multitensor version and get rid of this change to _multi_tensor_nadam I think I would have to remove line 226 of _single_tensor_nadam (`mu_product = mu_product * mu`), does that make sense?

Reply:
The two options I see are:
- We want to keep the exact same behavior as before and so we fold the second update for the single Tensor impl into the functional. And fold the multitensor computation into its functional version.
- We want to remove this duplicate code and so modify one of the two to match the other. In such case, I think we should keep the version that does not do the computation twice as it is better! Also the functional version should perform the full step, so I would argue it is a bug that part of it is done in the optimizer outside of the functional call.

Reply:
hm which option do you think I should go ahead with? It seems that ASGD and SGD have the same bug where part of the state is updated outside the functional form in single tensor ASGD and SGD as well. I think they did this (for ASGD and NAdam) because arguments were passed as floats rather than singleton tensors so they couldn't update state within the function.